### PR TITLE
[Frontend] re-enable multi-modality input in the new beam search implementation

### LINKED ISSUE #9577: [Usage]: beam_search not work with multimodal input
<details>
<summary>The output of `python collect_env.py`</summary>

```text
Collecting environment information...
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.31

Python version: 3.12.3 | packaged by Anaconda, Inc. | (main, May  6 2024, 19:46:43) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-71-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
[/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5)
[/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5](https://vscode-remote+ssh-002dremote-002b10-002e40-002e116-002e16.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          64
On-line CPU(s) list:             0-63
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD Ryzen Threadripper PRO 3975WX 32-Cores
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     3500.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        6986.53
Virtualization:                  AMD-V
L1d cache:                       1 MiB
L1i cache:                       1 MiB
L2 cache:                        16 MiB
L3 cache:                        128 MiB
NUMA node0 CPU(s):               0-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-ml-py==12.555.43
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.5.40
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] pytorch-lightning==2.3.0
[pip3] pyzmq==25.1.2
[pip3] torch==2.4.0
[pip3] torchmetrics==1.4.0.post0
[pip3] torchvision==0.19.0
[pip3] transformers==4.45.2
[pip3] triton==3.0.0
[pip3] tritonclient==2.48.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-ml-py              12.555.43                pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.5.40                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] pytorch-lightning         2.3.0                    pypi_0    pypi
[conda] pyzmq                     25.1.2          py312h6a678d5_0  
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchmetrics              1.4.0.post0              pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] transformers              4.45.2                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
[conda] tritonclient              2.48.0                   pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.3.post1
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0	GPU1	GPU2	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	PHB	SYS	0-63	0		N/A
GPU1	PHB	 X 	SYS	0-63	0		N/A
GPU2	SYS	SYS	 X 	0-63	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

</details>

### How would you like to use vllm

I'm trying to do batch inference with multimodal input data like this:

```
llm = LLM(model='OpenGVLab/InternVL2-1B', trust_remote_code=True, max_num_seqs=5, enforce_eager=True)
sampling_params = SamplingParams(max_tokens=128, temperature=0.0)
beam_params = BeamSearchParams(beam_width=3, max_tokens=128, temperature=0.0)
for image_file in image_files:
    input = {
              "prompt": prompt,
              "multi_modal_data": { "image": Image.open(image_file)},
            }
    list_input.append(input)
outputs = llm.generate(list_input, sampling_params)
outputs = llm.beam_search(list_input, beam_params)
```
The model run perfectly with generate() method but when I call beam_search() method I got this error:
TypeError: TextEncodeInput must be Union[TextInputSequence, Tuple[InputSequence, InputSequence]]

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by FerdinandZhong**
FIX #9577

**Changes in this PR:**

This PR introduces the following changes based on the updated beam search implementation:

* Re-enable multi-modality input:
Support for multi-modality input has been re-enabled for beam search with OpenAI-compatible endpoints.
* Logprobs handling in `ChatCompletionRequest`:
Added additional validation to disable l`ogprobs` when `use_beam_search=True`. Since the beam search selects results based on `cumulative logprobs` and determines step logprobs by `beam_width`, it ignores the `top_logprobs` and `logprobs` parameters passed in with the request.

**Unit Test**

Added two additional test cases in `tests/entrypoints/openai/test_vision.py`.

**Manual Testing**

The following command was used to launch the server for manual testing: `vllm serve microsoft/Phi-3.5-vision-instruct --api-key token-abc123 --trust-remote-code --max-model-len 4096 --limit-mm-per-prompt image=2`

Client script used to test the changes:
```python
import openai
import asyncio


url = "http://localhost:"
client = openai.AsyncOpenAI(
    base_url = "http://localhost:8000/v1",
    api_key="token-abc123"
)


# Image URLs
img_urls = [
    "https://upload.wikimedia.org/wikipedia/commons/c/cb/Brachiosaurus_DB_flipped.jpg",
    "https://upload.wikimedia.org/wikipedia/commons/3/3d/Allosaurus_Revised.jpg"
]

# Define the messages for the chat completion
messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image_url",
                "image_url": {
                    "url": img_urls[0]
                }
            },
            {
                "type": "image_url",
                "image_url": {
                    "url": img_urls[1]
                }
            },
            {
                "type": "text",
                "text": "what are the animals in the images?"
            }
        ]
    }
]

async def make_request():
    try:
        response = await client.chat.completions.create(
            model="microsoft/Phi-3.5-vision-instruct",
            max_tokens=32,
            temperature=0,
            messages=messages,
            n=2,
            extra_body={"use_beam_search": True}
        )
        for choice in response.choices:
            print(choice.message.content)

    except openai.BadRequestError as e:
        print(f"Error: {e.code}")

asyncio.run(make_request())
```
Verified the functionality of multi-image input handling and correct response generation using beam search with the above manual tests.

### CODE REVIEW

**Comment by FerdinandZhong** [comment]
Hi @simon-mo @khluu could you please add me to your Buildkite org to unblock the full CI run?

**Comment by simon-mo** [comment]
Added your email to our buildkite org.

File: vllm/engine/protocol.py
Comment (DarkLight1337) [inline vllm/engine/protocol.py:106]: Similar to the `Sequence` we use normally, let's place `multi_modal_data` and `mm_processor_kwargs` inside `BeamSearchSequence`.

File: vllm/beam_search.py
Comment (DarkLight1337) [inline vllm/beam_search.py:16]: Is this necessary?

File: vllm/engine/protocol.py
Comment (DarkLight1337) [inline vllm/engine/protocol.py:85]: We can use `vllm.inputs.parse_singleton_prompt` here.

File: vllm/engine/protocol.py
Comment (DarkLight1337) [inline vllm/engine/protocol.py:93]: We can use `InputPreprocessor._extract_prompt_components` here.

File: vllm/beam_search.py
Comment (FerdinandZhong) [reply vllm/beam_search.py:19]: Hi, thank you for the comment.
Yes I think so. Besides the finish_reason I also added the `stop_reason` in my latest commit to this class. These two values are passed to `CompletionOutput` when the generation is done

File: vllm/engine/protocol.py
Comment (FerdinandZhong) [reply vllm/engine/protocol.py]: Thank you for the comment, I added `multi_modal_data` and `mm_processor_kwargs` to the class

File: vllm/engine/protocol.py
Comment (FerdinandZhong) [reply vllm/engine/protocol.py]: Thank you! I changed to this function

File: vllm/engine/protocol.py
Comment (DarkLight1337) [inline vllm/engine/protocol.py:78]: Btw why is this function defined inside a protocol? Perhaps we should move this to `LLMEngine`? Then we can make use of the existing `input_preprocessor` defined there. @youkaichao

Comment (FerdinandZhong) [reply vllm/engine/protocol.py]: Sorry I made a mistake here, I just pushed another commit to change it from `self.input_preprocessor` and `self.tokenizer` to `input_preprocessor` and `tokenizer`.

By the way, in the current release v0.6.3, this function is inside `AsyncLLMEngine` and it has been moved here in a recent pr: [9296](https://github.com/vllm-project/vllm/pull/9296)

**Comment by FerdinandZhong** [comment]
Hi @DarkLight1337, thank you for the review!

I noticed the changes in PR [#9473](https://github.com/vllm-project/vllm/pull/9473) and will merge the latest code once that PR is merged. Regarding the conflicts between the two PRs:

* Logprobs: I'll align with the fix in [#9473](https://github.com/vllm-project/vllm/pull/9473). In my PR, I directly prevent the return of logprobs when using beam search, as the number of logprobs in each step is determined by beam_width.

* PromptType: I've kept it as the input type for handling multi-modality data. To correctly process the prompt passed to the function, I use `inputpreprocess` to parse the content, as suggested by @DarkLight1337. Additionaly, I also have `prompt_text` (parsed from the prompt) set as RequestOutput.prompt, which resolves the error related to `PromptType` being the input.

@njhill, could you please review these changes?

**Comment by njhill** [comment]
Thanks @FerdinandZhong, I'll review tomorrow (Friday US time)

**Comment by FerdinandZhong** [comment]
> Thanks @FerdinandZhong, I'll review tomorrow (Friday US time)

Thanks @njhill. May I know if the change looks good to you?

**Comment by njhill** [comment]
@FerdinandZhong sorry for the delay, bit behind with things.

The changes look good to me thanks. The InputProcessor change makes sense.

Re the logprobs, it's a good point that the number returned will be based on the beam width rather than how many are actually requested. I think we can improve this to request the max of these two and truncate as needed. But no need to change that for this PR.

I think we can improve the impl quite a bit more overall in some follow-on updates including:
- Support most/all params.. I don't see any reason we can't, this should actually be easier with the "external" impl
- Remove the separate beam search API, we can retain the function of the existing beam_search parameters, just have this layer intercept those
- Move the impl out of protocol.py .. at a minimum we can have it in beam_search.py and just call it from the abstract base class

**Review by njhill — APPROVED** [review]

**Comment by FerdinandZhong** [comment]
> @FerdinandZhong sorry for the delay, bit behind with things.
> 
> The changes look good to me thanks. The InputProcessor change makes sense.
> 
> Re the logprobs, it's a good point that the number returned will be based on the beam width rather than how many are actually requested. I think we can improve this to request the max of these two and truncate as needed. But no need to change that for this PR.
> 
> I think we can improve the impl quite a bit more overall in some follow-on updates including:
> 
> * Support most/all params.. I don't see any reason we can't, this should actually be easier with the "external" impl
> * Remove the separate beam search API, we can retain the function of the existing beam_search parameters, just have this layer intercept those. I'm not sure that we actually need separate BeamSearchParams.
> * Move the impl out of protocol.py .. at a minimum we can have it in beam_search.py and just call it from the abstract base class

Hi @njhill, thank you for your comment!

I agree that taking the maximum of the `beam_width` and `top_logprobs` can be a good idea, and we can implement that change in the following PR. I'm also aligned with the action points you mentioned for improving beam search, and I'd be happy to collaborate on these enhancements moving forward.

**Comment by youkaichao** [comment]
I'm surprised there are so many efforts for adding various features for beam-search ...

We will work for implementing beam search in another way so that all features for normal generation just works.

**Comment by FerdinandZhong** [comment]
Hi @youkaichao, thank you for the feedback. I understand your concerns, and I agree that, in the long run, beam search can be properly designed and implemented. In the short term, I’m happy to continue providing feedback and contributing commits from a user’s perspective.

In the meantime, @DarkLight1337, could we consider merging this PR first, as it addresses the fix for [#9577](https://github.com/vllm-project/vllm/issues/9577)?

**Comment by DarkLight1337** [comment]
@njhill do we need to rebuild this? Seem that you cancelled some of the tests.

**Comment by FerdinandZhong** [comment]
Hi @DarkLight1337 , I’ve merged the latest main branch and rerun the tests from my end. May I ask for your advice on the next steps to do with this PR for merging? Additionally, can I check with you if rebase is needed to add "sign-off" for each commit? Thank you.

**Comment by DarkLight1337** [comment]
Sorry for missing this! You don't have to worry about signing-off for this PR as we can manually set that to pass.

**Comment by DarkLight1337** [comment]
I have enabled auto-merge which should run all the tests and merge if they pass.

**Comment by FerdinandZhong** [comment]
@DarkLight1337 got it, thanks!

**Pull request merged**
